File: algorithm/dfs/cpp/number_of_provinces.cpp
// https://leetcode.com/problems/number-of-provinces

/*
There are n cities. Some of them are connected, while some are not. If city a is connected directly with city b, and city b is connected directly with city c, then city a is connected indirectly with city c.
A province is a group of directly or indirectly connected cities and no other cities outside of the group.
You are given an n x n matrix isConnected where isConnected[i][j] = 1 if the ith city and the jth city are directly connected, and isConnected[i][j] = 0 otherwise.
Return the total number of provinces.

Ex1:
Input: isConnected = [[1,1,0],[1,1,0],[0,0,1]]
Output: 2

Ex2:
Input: isConnected = [[1,0,0],[0,1,0],[0,0,1]]
Output: 3

*/

#include <vector>
#include <cassert>
#include <unordered_set>
#include <unordered_map>

using namespace std;

// Time: O(N^2), we need n^2 time to construct graph, and also dense graph can have N^2 number of edges
// Space: O(N^2)
void dfs(int curr, unordered_map<int, vector<int>>& graph, unordered_set<int>& visited) {
    for (int next : graph[curr]) {
        if (visited.count(next)) continue;
        visited.insert(next);
        dfs(next, graph, visited);
    }
}

int findCircleNum(vector<vector<int>>& isConnected) {
    if (isConnected.empty() || isConnected[0].empty()) {
        return 0;
    }

    unordered_map<int, vector<int>> graph;
    
    int n = isConnected.size();
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            if (isConnected[i][j] == 1) {
                graph[i].push_back(j);
                graph[j].push_back(i);
            }
        }
    }

    int res = 0;
    unordered_set<int> visited;
    for (int i = 0; i < n; i++) {
        if (visited.count(i)) continue;
        visited.insert(i);
        dfs(i, graph, visited);
        res++;
    }

    return res;
}

// Time: O(N^2), Space: O(N)
void dfs(vector<vector<int>>& isConnected, int curr, unordered_set<int>& visited) {
    for (int i = 0; i < isConnected[curr].size(); i++) {
        int neighbor = i;
        if (visited.count(neighbor) || isConnected[curr][neighbor] == 0) continue;
        visited.insert(neighbor);
        dfs(isConnected, neighbor, visited);
    }
}

int findCircleNum(vector<vector<int>>& isConnected) {
    int n = isConnected.size();
    int res = 0;
    unordered_set<int> visited;

    for (int i = 0; i < n; i++) {
        if (visited.count(i)) {
            continue;
        }
        visited.insert(i);
        dfs(isConnected, i, visited);
        res++;
    }

    return res;
}

int main() {
    vector<vector<int>> isConnected1 = {{1,1,0},{1,1,0},{0,0,1}};
    assert(findCircleNum(isConnected1) == 2);

    vector<vector<int>> isConnected2 = {{1,0,0},{0,1,0},{0,0,1}};
    assert(findCircleNum(isConnected2) == 3);

    return 0;
}

File: algorithm/dfs/python/number_of_provinces.py
"""
Number of Provinces
https://leetcode.com/problems/number-of-provinces

Approaches:
1. Build an adjacency-list graph, then DFS from each unvisited node.
   Time: O(N^2) to build the graph (dense) and DFS. Space: O(N^2).
2. DFS directly on the adjacency matrix without building a separate graph.
   Time: O(N^2). Space: O(N).
"""

from collections import defaultdict
from typing import Dict, List, Set


class Solution:
    # Time: O(N^2), we need N^2 time to construct graph, and a dense graph can have N^2 edges.
    # Space: O(N^2)
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        if not isConnected or not isConnected[0]:
            return 0

        graph: Dict[int, List[int]] = defaultdict(list)

        n = len(isConnected)
        for i in range(n):
            for j in range(n):
                if isConnected[i][j] == 1:
                    graph[i].append(j)
                    graph[j].append(i)

        res = 0
        visited: Set[int] = set()
        for i in range(n):
            if i in visited:
                continue
            visited.add(i)
            self._dfs(i, graph, visited)
            res += 1

        return res

    def _dfs(self, curr: int, graph: Dict[int, List[int]], visited: Set[int]) -> None:
        for nxt in graph[curr]:
            if nxt in visited:
                continue
            visited.add(nxt)
            self._dfs(nxt, graph, visited)


class Solution2:
    # Time: O(N^2), Space: O(N)
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        n = len(isConnected)
        res = 0
        visited: Set[int] = set()

        for i in range(n):
            if i in visited:
                continue
            visited.add(i)
            self._dfs(isConnected, i, visited)
            res += 1

        return res

    def _dfs(self, isConnected: List[List[int]], curr: int, visited: Set[int]) -> None:
        for i in range(len(isConnected[curr])):
            neighbor = i
            if neighbor in visited or isConnected[curr][neighbor] == 0:
                continue
            visited.add(neighbor)
            self._dfs(isConnected, neighbor, visited)


if __name__ == "__main__":
    isConnected1 = [[1, 1, 0], [1, 1, 0], [0, 0, 1]]
    assert Solution().findCircleNum(isConnected1) == 2
    assert Solution2().findCircleNum(isConnected1) == 2

    isConnected2 = [[1, 0, 0], [0, 1, 0], [0, 0, 1]]
    assert Solution().findCircleNum(isConnected2) == 3
    assert Solution2().findCircleNum(isConnected2) == 3
